File: arrays_and_hashing/Two_sums.ts
/*
 * PROBLEM: Two sums
 * LINK: https://neetcode.io/problems/two-integer-sum
 * -------------------------------------------------------------------------
 * APPROACH EVOLUTION:
 * 1. Brute Force: O(n^2) - Nested loops
 * -------------------------------------------------------------------------
 * TIME COMPLEXITY:  O(n^2) - Single pass each array number
 * SPACE COMPLEXITY: O(1) -
 * PERFORMANCE: Beats 74.55% in Runtime
 * -------------------------------------------------------------------------
 */
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @return {number[]}
     */
    twoSum(nums: [], target: number) {
        for(let i = 0;i < nums.length;i++){
            for(let j = i + 1;j < nums.length;j++){
                if(nums[i] + nums[j] === target){
                    return [i, j];
                }
            }
        }
    }
}

File: arrays_and_hashing/Two_sums.go
/*
 * PROBLEM: Two sums
 * LINK: https://neetcode.io/problems/two-integer-sum
 * -------------------------------------------------------------------------
 * APPROACH EVOLUTION:
 * 1. Optimized: O(n) - Map one pass
 * -------------------------------------------------------------------------
 * TIME COMPLEXITY:  O(n) - Constant time loop with Map
 * SPACE COMPLEXITY: O(n) - Storing up to N elements in the Map
 * -------------------------------------------------------------------------
 */
package main

import "fmt"

func twoSum(nums []int, target int) []int {
	numbers := make(map[int]int)

	for i, num := range nums {
		completeNumber := target - num

		if index, ok := numbers[completeNumber]; ok {
			return []int{index, i}
		}

		numbers[num] = i
	}

	return []int{}
}

func mainTwoSums() {
	fmt.Println(twoSum([]int{3, 4, 5, 6}, 7))
}
